### ISSUE TITLE
Was unable to create branch or PR - details for changes for the fixes below

### ISSUE THREAD

**Issue by macnak** [gaurishhs/elysia-session#3]
Hi,

I tried to branch and upload the changes make PR, no access. All good, below are the changes I made to get this working for more than two requests in a session. The issue was originally `let sid: string = ''` and if there already was a session cookie the sid value did not get set to that after the `if (cookie) {` check. So, the session cookie was empty value on the response to the second request. And then removing the two `undefined` parameters on the store functions.

index.ts
```ts
import type Elysia from "elysia";
import { Store } from "./store";
import { Session, SessionData } from "./session";
import { nanoid } from "nanoid";
import { CookieStore } from "./stores/cookie";
import { CookieOptions } from "elysia/dist/cookie";

export interface SessionOptions {
    store: Store;
    expireAfter: number;
    cookieName?: string;
    cookieOptions?: CookieOptions
}

export const sessionPlugin = (options: SessionOptions) => (app: Elysia) => {
    return app.derive(async (ctx) => {
        const store = options.store
        const session = new Session
        const cookieName = options.cookieName || 'session'
        const cookie = ctx.cookie[cookieName]
        let sid: string = '';
        let sessionData: SessionData | null | undefined
        let createRequired = false

        if (cookie) {
            sid = cookie.value;
            try {
                sessionData = await store.getSession(sid, ctx)
            } catch {
                createRequired = true
            }

            if (sessionData) {
                session.setCache(sessionData)

                if (session.valid()) {
                    session.reUpdate(options.expireAfter)
                } else {
                    store.deleteSession(sid, ctx)
                    createRequired = true
                }
            } else {
                createRequired = true
            }
        } else {
            createRequired = true
        }

        if (createRequired) {
            const initialData = {
                _data: {},
                _expire: null,
                _delete: false,
                _accessed: null
            }
            sid = nanoid(24)
            store.createSession(initialData, sid, ctx)
            session.setCache(initialData)
        }

        if (!(store instanceof CookieStore)) {
            ctx.cookie[cookieName].set({
                value: sid,
                ...options.cookieOptions
            })
        }

        store.persistSession(session.getCache(), sid, ctx)

        if (session.getCache()._delete) {
            store.deleteSession(sid, ctx)
        }

        return {
            session
        }
    })
}
```

As Mongoose uses Promises I made a mistake in my original code of not catering for this, head desk moment when I ran this under load and saw my mistake. So, change the interface functions to async and used awaits. This allowed the `getSession` function do the right thing and only return the SessionData object. And people can extend the ISession to add more of their own data that will go down into the persistent store. Further to the Schema parts the createdAt / modifiedAt properties flag can be set. So cleanup can be done automatically when the collection has the Schema attached to the model. I did not implement this, more a note for anyone who may use this to be aware that old sessions can build up in the store (mongodb) and that trimming old session records is a good idea.

Then for the MongooseStore example:
```ts
import { Context } from "elysia";
import { SessionData } from "elysia-session/dist/session";
import { Store } from "elysia-session/dist/store";
import * as mongoose from 'mongoose';



export interface ISession extends mongoose.Document {
  _id: string;
  sessionData: SessionData;
}


export class MongooseStore implements Store {
  private db: typeof import('mongoose');
  private collection: string;
  private schema: mongoose.Schema | null;
  private model: mongoose.Model<ISession> | null;

  constructor(db: typeof import('mongoose'), collection: string) {
    this.db = db;
    this.collection = collection;
    this.schema = new mongoose.Schema({
      _id: String,
      sessionData: { type: JSON },
    })
    this.model = mongoose.model<ISession>(this.collection, this.schema);
  }

  async getSession (id?: string | undefined, ctx?: Context): Promise<any> {
    if (!id) return null;
    if (this.model) {
      const value = await this.model.findOne({ _id: id }, { _id: 0, sessionData: 1 });
      if (!value) return null;
      return value.sessionData;
    } else
      return null
  }

  async createSession (data: SessionData, id: string, ctx?: Context): Promise<any> {
    if (this.model) {
      const session = new this.model({
        _id: id,
        sessionData: data,
      })
      await session.save();
    }
  }

  async deleteSession (id?: string | undefined, ctx?: Context): Promise<any> {
    if (!id) return;
    if (this.model) {
      await this.model.deleteOne({ _id: id })
    }
  }

  async persistSession (data: SessionData, id?: string, ctx?: Context): Promise<any> {
    if (!id) return;
    if (this.model) {
      await this.model.updateOne({ _id: id }, { sessionData: data })
    }
  }
}
```

**Comment by gaurishhs** [comment]
I don't think so we should be setting sid if cookie is present should be
set when sessionData is there

**Comment by gaurishhs** [comment]
Also you probably can't remove undefined since most stores don't need that
Param but some do

**Comment by macnak** [comment]
Okay, I've looked into this more. One core issue I found was that 1st request gets the cookie with the side - so set-cookie on the response, 2nd request has the cookie, response has the cookie but the side (value) is empty (when using any of the stores buy the cookie store as that uses the ctx).

And working out why, the answer is that MongooseStore is based off how the sqllite / memory store code, which uses the sid for id to resolve the SessionData in the store and not the ctx. While the cookie store used the ctx. So the cookie store will work correctly all the time as the store.getSession(undefined, ctx) will always have the cookie value (which is the sid) and from that will return the SessionData as the session cookie value is the SessionData

You still have the issue with sid being empty string and never set so when the code below is expected (on the second request for the session) that means on the response header the set-cookie will have the cookie with empty value as sid is empty string (This was the issue I ran into when more than 2 requests on the same session cookie had been sent. The response on the second request sent empty cookie value and session functionality stopped working at that point:

```ts
        if (!(store instanceof CookieStore)) {
            ctx.cookie[cookieName].set({
                value: sid,
                ...options.cookieOptions
            })
        }
```

The SessionData will never be returned for the sqllite / memory stores as it uses the id which is always undefined and the ctx is never used. Why as they used the id parameter on the getSession API call which is set to undefined and the ctx is not used like cookie store does.

The odd store out is cookie store which does not care about the sid, while the other do care about sid as that is the key used to get the SessionData, this is why you have for the id either string | undefined. To solve that make a change earlier in the code to set the sid if the cookie is present instead of the default of empty string:

```ts
        let sid: string = cookie === undefined ? undefined : cookie.value;
```

this means sid will be undefined if the cookie does not exist, else it will be the sid if using the none Cookie store, or the SessionData if this is using the CookieStore. Now sid can be passed without issue to the two store calls "getSession() | deleteSession()". It really does not matter for Cookie Store as it does not care about the sid value as it is not used anyway, so undefine or string - no difference in the function operation.

Another thing is doing the check "session.valid()" in index.ts - I think that is a bad idea to do here. I would document for people using this to do the check in their Web API end points. With what has been done in this library they will never know if the session was invalid as a new one is created automatically when the current is detected as invalid due to it expiring.
